test/controllers/supplementary/TransportInformationContainersPageControllerSpec.scala

```diff
@@ -0,0 +1,269 @@
+      }
+
+      "try to save more than 9999 containers" in {
+        withCaching[TransportInformationContainerData](Some(cacheWithMaximumAmountOfHolders), id)
+
+        val body = Seq(("id", "M1l3s"), saveAndContinueActionURLEncoded)
+        val result = route(app, postRequestFormUrlEncoded(uri, body: _*)).get
+
+        status(result) must be(BAD_REQUEST)
+        contentAsString(result) must include(messages("supplementary.limit"))
+      }
+
+      "try to remove a non existent container" in {
+        val cachedData = TransportInformationContainerData(Seq(TransportInformationContainer("M1l3s")))
+        withCaching[TransportInformationContainerData](Some(cachedData), id)
+
+        val body = ("action", "Remove:J0ohn-Coltrane")
+
+        val result = route(app, postRequestFormUrlEncoded(uri, body)).get
+        val stringResult = contentAsString(result)
+
+        status(result) must be(BAD_REQUEST)
+        stringResult must include(messages("global.error.title"))
+        stringResult must include(messages("global.error.heading"))
+        stringResult must include(messages("global.error.message"))
+      }
+    }
+  }
+  "redirect to the next page" when {
+
+    "user provide container with empty cache" in {
+
+      val body = Seq(("id", "M1l3s"), saveAndContinueActionURLEncoded)
+
+      val result = route(app, postRequestFormUrlEncoded(uri, body: _*)).get
+      val header = result.futureValue.header
+
+      status(result) must be(SEE_OTHER)
+      header.headers.get("Location") must be(
+        Some("/customs-declare-exports/declaration/supplementary/total-numbers-of-items")
+      )
+    }
+
+    "user doesn't fill form but some containers already exist in the cache" in {
+      val cachedData = TransportInformationContainerData(Seq(TransportInformationContainer("Jo0hn")))
+      withCaching[TransportInformationContainerData](Some(cachedData), id)
+
+      val result = route(app, postRequestFormUrlEncoded(uri, saveAndContinueActionURLEncoded)).get
+      val header = result.futureValue.header
+
+      status(result) must be(SEE_OTHER)
+      header.headers.get("Location") must be(
+        Some("/customs-declare-exports/declaration/supplementary/total-numbers-of-items")
+      )
+    }
+
+    "user provide container with some different container in cache" in {
+      val cachedData = TransportInformationContainerData(Seq(TransportInformationContainer("x4rlz")))
+      withCaching[TransportInformationContainerData](Some(cachedData), id)
+
+      val body = Seq(("id", "M1l3s"), saveAndContinueActionURLEncoded)
+
+      val result = route(app, postRequestFormUrlEncoded(uri, body: _*)).get
+      val header = result.futureValue.header
+
+      status(result) must be(SEE_OTHER)
+      header.headers.get("Location") must be(
+        Some("/customs-declare-exports/declaration/supplementary/total-numbers-of-items")
+      )
+    }
+  }
+}
+
+object TransportInformationContainersPageControllerSpec {
```

Comment:
why do you need an object creation in here?

Reply:
Companion object is a good place to store some constant data. We started using it for tests, so whole constants you will have in one place.

Reply:
might be worth companion objects for tests at one place we will to be aware of the re-use of they are really created for that purpose.

Reply:
If they live in the same class then we should be able to check if some has been created already.

Reply:
Companion object should be close to the class. We shouldn't move it from this file. It's not an object to use by all classes. It must contain only stuff for this test.